pkg/kubectl/cmd/apply.go

```diff
@@ -0,0 +1,161 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/spf13/cobra"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/kubectl"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+	"k8s.io/kubernetes/pkg/util/strategicpatch"
+)
+
+// ApplyOptions stores cmd.Flag values for apply.  As new fields are added, add them here instead of
+// referencing the cmd.Flags()
+type ApplyOptions struct {
+	Filenames []string
+}
+
+const (
+	apply_long = `Apply a configuration to a resource by filename or stdin.
+
+JSON and YAML formats are accepted.`
+	apply_example = `# Apply the configuration in pod.json to a pod.
+$ kubectl apply -f ./pod.json
+
+# Apply the JSON passed into stdin to a pod.
+$ cat pod.json | kubectl apply -f -`
+)
+
+func NewCmdApply(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+	options := &ApplyOptions{}
+
+	cmd := &cobra.Command{
+		Use:     "apply -f FILENAME",
+		Short:   "Apply a configuration to a resource by filename or stdin",
+		Long:    apply_long,
+		Example: apply_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(validateArgs(cmd, args))
+			cmdutil.CheckErr(cmdutil.ValidateOutputArgs(cmd))
+			cmdutil.CheckErr(RunApply(f, cmd, out, options))
+		},
+	}
+
+	usage := "Filename, directory, or URL to file that contains the configuration to apply"
+	kubectl.AddJsonFilenameFlag(cmd, &options.Filenames, usage)
+	cmd.MarkFlagRequired("filename")
+	cmdutil.AddValidateFlags(cmd)
+	cmdutil.AddOutputFlagsForMutation(cmd)
+	return cmd
+}
+
+func validateArgs(cmd *cobra.Command, args []string) error {
+	if len(args) != 0 {
+		return cmdutil.UsageError(cmd, "Unexpected args: %v", args)
+	}
+
+	return nil
+}
+
+func RunApply(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, options *ApplyOptions) error {
+	shortOutput := cmdutil.GetFlagString(cmd, "output") == "name"
+	schema, err := f.Validator(cmdutil.GetFlagBool(cmd, "validate"), cmdutil.GetFlagString(cmd, "schema-cache-dir"))
+	if err != nil {
+		return err
+	}
+
+	cmdNamespace, enforceNamespace, err := f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+
+	mapper, typer := f.Object()
+	r := resource.NewBuilder(mapper, typer, f.ClientMapperForCommand()).
+		Schema(schema).
+		ContinueOnError().
+		NamespaceParam(cmdNamespace).DefaultNamespace().
+		FilenameParam(enforceNamespace, options.Filenames...).
+		Flatten().
+		Latest().
+		Do()
+	err = r.Err()
+	if err != nil {
+		return err
+	}
+
+	count := 0
+	err = r.Visit(func(info *resource.Info, err error) error {
+		// In this method, info.Object contains the object retrieved from the server
+		// and info.VersionedObject contains the object decoded from the input source.
+		if err != nil {
+			return err
+		}
+
+		// Serialize the current configuration of the object from the server.
+		current, err := info.Mapping.Codec.Encode(info.Object)
```

Comment:
Per above, remove `Latest()` and change the order you serialize `current`, `original`, `modified` to
```
// Do this BEFORE calling info.Get()
modified, err := kubectl.GetModifiedConfiguration(info, true)
if err != nil {
		...
}

// Now call info.Get() to retrieve the latest version
if err := info.Get(); err != nil {
   ...
}

current, err := info.Mapping.Codec.Encode(info.Object)
if err != nil {
		...
}

// Now, retrieve original configuration from the latest server-side version of the object
original, err := kubectl.GetOriginalConfiguration(info)
if err != nil {
		...
}

// Compute a three way strategic merge patch ...
```